rollout/trafficrouting.go

```diff
@@ -126,6 +126,15 @@ func (c *rolloutContext) reconcileTrafficRouting() error {
 			if c.rollout.Spec.Strategy.Canary.DynamicStableScale {
 				desiredWeight = 100 - ((100 * c.stableRS.Status.AvailableReplicas) / *c.rollout.Spec.Replicas)
 			}
+		} else if c.rollout.Status.PromoteFull {
+			// on a promote full, desired stable weight should be 0 (100% to canary),
+			// But we can only increase canary weight according to available replica counts of the canary.
+			// we will need to set the desiredWeight to 0 when the newRS is not available.
+			newRSAvailableReplicas := int32(0)
+			if c.newRS != nil {
+				newRSAvailableReplicas = c.newRS.Status.AvailableReplicas
+			}
+			desiredWeight = (100 * newRSAvailableReplicas) / *c.rollout.Spec.Replicas
```

Comment:
When promote full happens for canary with traffic routing, the desired behavior is we go from the current percentage to 100% (instead of a gradual). This is to support use case for users who are using the canary strategy like a blue-green, going from 0% to 100%, because their API cannot tolerate two versions. Unfortunately, this technique will break that experience so I think we will need to find a different way to fix the bug.

Reply:
> When promote full happens for canary with traffic routing, the desired behavior is we go from the current percentage to 100% (instead of a gradual). This is to support use case for users who are using the canary strategy like a blue-green, going from 0% to 100%, because their API cannot tolerate two versions. Unfortunately, this technique will break that experience so I think we will need to find a different way to fix the bug.

Users who are using the canary strategy like a blue-green it doesn't make sense for them to use the `dynamicStableScale` feature,  as before promotion we will need the canary and stable to be 100% scaled until the rollout is fully promoted in order that when we do cut off the traffic we will not have an outage.

Therefore my suggestion is 

- if  `dynamicStableScale` is `true` then we will go with the dynamic traffic routing. 
- if `dynamicStableScale` is `false`  then leave the current behavior, the canary and stable will be 100% scaled and the traffic will go from 0% to 100%.

@jessesuen does this seem reasonable?

Reply:
If someone wants a blue-green deployment then shouldn't they be using the blue-green strategy?  If we're going from 50% to 100%, it makes sense to get as much traffic on the new version as we can (given pod readiness)

Reply:
According to https://github.com/argoproj/argo-rollouts/issues/1677#issuecomment-985796488 - `promote -full` is a break-glass operation that should shift all the traffic no matter what.
This change implies that there is only a subsection of the whole traffic sent to the new canary (or maybe i'm misreading the code)?

Reply:
> According to #1677 (comment) - promote -full is a break-glass operation that should shift all the traffic no matter what.
This change implies that there is only a subsection of the whole traffic sent to the new canary (or maybe i'm misreading the code)?

Correct, the traffic will be shifted by the percentage of the available canary pods until the new canary pods will be 100% scaled then it will get 100% of the traffic, this guarantees that the new canary pods are scaled enough to handle the traffic.

Reply:
> Therefore my suggestion is
> if dynamicStableScale is true then we will go with the dynamic traffic routing.
> if dynamicStableScale is false then leave the current behavior, the canary and stable will be 100% scaled and the traffic will go from 0% to 100%.

@nissy34 Does the bug only happen with `dynamicStableScale: true`? If so, I agree with your suggestion. In fact, your suggestion is something I considered myself but wasn't sure if we were still leaving a bug open for `dynamicStableScale: false` users. 

> If someone wants a blue-green deployment then shouldn't they be using the blue-green strategy? If we're going from 50% to 100%, it makes sense to get as much traffic on the new version as we can (given pod readiness)

There are valid reasons to accomplish blue-green using mesh/ingress provider features rather than service level cutovers because there are implications for changing the service selectors. 

For example, when speaking with AWS team, they recommended to us that blue-green be achieved using weighted target groups instead of service selector cutovers when using AWS load balancer controller. This is because service label modifications do not allow pod readiness gates to be injected properly (since aws-load-balancer-controller relies on service labels to determine the correct pods to inject).

Another reason is, when we have more advanced traffic shaping features like shadow or header-based traffic shaping, users will still want to use the canary strategy instead of blue-green, because the blue-green strategy won't be able to shadow or canary traffic based on header, while still doing a 0% to 100% cutover for production users.

Reply:
> Does the bug only happen with `dynamicStableScale: true`?

I played around with `dynamicStableScale: true` and the bug is still present.

The root cause is when it takes more than a single reconciliation until the canary replica sets are available

As at the end of the first reconcile the currentStepIndex is set to equal the number of steps  
https://github.com/argoproj/argo-rollouts/blob/master/rollout/canary.go#L353-L359
```go
	if c.rollout.Status.PromoteFull {
		c.pauseContext.ClearPauseConditions()
		c.pauseContext.RemoveAbort()
		if stepCount > 0 {
			currentStepIndex = &stepCount
		}
	}
```
And in the next reconciliation in reconcileTrafficRouting
https://github.com/argoproj/argo-rollouts/blob/master/rollout/trafficrouting.go#L134-L153

because that `!atDesiredReplicaCount && !c.rollout.Status.PromoteFull` is `false` and `*index != int32(len(c.rollout.Spec.Strategy.Canary.Steps))` is also `false` because in the prev reconcile the currentStepIndex  was set to `int32(len(c.rollout.Spec.Strategy.Canary.Steps))`

So the desiredWeight will be 0 even so we didn't yet promote the newRS to be stable ( as not all pods are ready yet).

Can we use ` c.rollout.Status.Canary.Weights.Canary.Weight` to Freeze the weigts?
if so we can do in https://github.com/argoproj/argo-rollouts/blob/master/rollout/trafficrouting.go#L132
```go
if c.rollout.Status.PromoteFull {
	// on a promote full, desired stable weight should be 0 (100% to canary),
	// But we can only increase canary weight according to available replica counts of the canary.
	// we will need to set the desiredWeight to 0 when the newRS is not available.
	if c.rollout.Spec.Strategy.Canary.DynamicStableScale {
		desiredWeight = (100 * c.newRS.Status.AvailableReplicas) / *c.rollout.Spec.Replicas
	} else {
		desiredWeight = c.rollout.Status.Canary.Weights.Canary.Weight
	}
}
```